### PR TITLE
Deprecate cert gen without a CA and add self-signed option

### PR DESCRIPTION
**Pull request by ywangd**
Generating a CA on the fly is an attempt at workflow optimisation that was inherited from certgen. There are potential pitfalls with this approach. Overall it is recommended to separate the step of CA creation and mandate a CA to be specified when generating certificate.

This PR add a deprecation message if the cert command is used without specifying a CA. A follow up PR will throw error for this usage in 8.0.

For use case where we explicitly trust a certificate without needing a CA, e.g. SAML message signing, the PR adds a `--self-signed` option to the `cert` sub-command to generate self-signed certificate.

Relates: #61884

### CODE REVIEW

**Comment by ywangd** [comment]
Doc change preview: https://elasticsearch_64037.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/certutil.html#certutil-cert

**Comment by lcawl** [comment]
Does the example in this section of the SAML guide need to be updated to use a CA too?: https://www.elastic.co/guide/en/elasticsearch/reference/master/saml-guide-authentication.html#_generating_certificates_and_keys

**Review by lcawl — APPROVED** [review]
One comment, otherwise LGTM

**Comment by jkakavas** [comment]
> Does the example in this section of the SAML guide need to be updated to use a CA too?: https://www.elastic.co/guide/en/elasticsearch/reference/master/saml-guide-authentication.html#_generating_certificates_and_keys

In SAML we don't use PKI but explicit trust of certificates so keeping the CA keys around is not necessary. This was actually one of the cases where the behavior we are now deprecating and removing was helpful and made sense. Instead of doing a 2 step process just for the sake of using `certutil`here, could we have an example with `openssl` ?

**Comment by tvernum** [comment]
Is it worth adding a `--self-signed` option to `certutil cert` to cater for the SAML use case?

**Comment by ywangd** [comment]
> This was actually one of the cases where the behavior we are now deprecating and removing was helpful and made sense. Instead of doing a 2 step process just for the sake of using certutilhere, could we have an example with openssl ?
>
Thakns for this info. For me, replacing it with openssl has a few downsides:
* Introduces another dependency. IIUC, elasticsearch-certutil was created to (at least partially) avoid having to use `openssl`. So would it feel backwards if we have to turn back to openssl?
* Most other places (if not all) uses openssl, so it is inconsistent.
* As I understand it, openssl has some inconsistency/installation issues across different OS (though it may not be an issue for the `req` command that we need here).
* The existing saml-guide doc page needs some more wording changes to accomodate the openssl replacement 

> Is it worth adding a --self-signed option to certutil cert to cater for the SAML use case?
> 
This is plausible when you need only a single cert. However, it could be problematic when creating multiple certificates. If users try to use it this for multiple nodes, it will at least cause confusion. We could forbid generating mulitple certs if `--self-signed` is requested. But even with that, it could still cause confusion in a TLS context, since our docs differentiate between cert and CA most of the times (if not all the time). There is also a minor difference between using two calls (`ca` and `cert`, or `cert` with on the fly CA) and self-signed certificate. The self-signed cert has `TRUE` value for the CA attribute while the other one has it as `FALSE`. I don't think it really matter in our use case. Just bring it up in case there are unobvious issues.

So given the above information, I wonder whether we should re-evaluate our decision on this issue, i.e. whether we should disallow generating CA on the fly entirely? By disallowing it, two-step process is always required to generate CA and certs. Does it go against "easier security setup" idea that we try to uphold? Or will "security-on-by-default" provide a different solution to make this worry obsolete?

PS: Another place we generate CA and certs together is for [configuring TLS in docker](https://www.elastic.co/guide/en/elasticsearch/reference/current/configuring-tls-docker.html#CO523-1).

**Comment by jkakavas** [comment]
> Is it worth adding a --self-signed option to certutil cert to cater for the SAML use case?

This sounds like a good idea to me. 

> Introduces another dependency. IIUC, elasticsearch-certutil was created to (at least partially) avoid having to use openssl. So would it feel backwards if we have to turn back to openssl?

I don't see this as a dependency, but merely as a suggestion as to how one can generate a keypair (more?) easily if openssl is available. We can have instructions with certutil (only|too), it's just two commands instead of one. 

> The existing saml-guide doc page needs some more wording changes to accomodate the openssl replacement

Making changes is not a downside in my opinion. 

> As I understand it, openssl has some inconsistency/installation issues across different OS (though it may not be an issue for the req command that we need here).

I don't think there is any problematic behavior wrt to key/certificate generation ( especially for certificates that are merely public key containers )

> However, it could be problematic when creating multiple certificates. If users try to use it this for multiple nodes, it will at least cause confusion

I think there is _that much_ we can do to prevent misuse of the tools we create. In the same line of thinking, nothing prevents a user to generate a new CA and cert on each node if they are not familiar enough with PKI and certificates. 

> . But even with that, it could still cause confusion in a TLS context, since our docs differentiate between cert and CA most of the times (if not all the time)

Can you explain your concern?

> The self-signed cert has TRUE value for the CA attribute while the other one has it as FALSE. I don't think it really matter in our use case

+1

> So given the above information, I wonder whether we should re-evaluate our decision on this issue, i.e. whether we should disallow generating CA on the fly entirely? By disallowing it, two-step process is always required to generate CA and certs. Does it go against "easier security setup" idea that we try to uphold? Or will "security-on-by-default" provide a different solution to make this worry obsolete?

I don't think we should re-evaluate.  Security on by default would probably not replace our existing tooling with other solutions and I don't see this as making it more difficult for users to set up TLS. The SAML use case affects a fraction of our users since a) request signing is not enabled by default and b) response encryption is not commonly used. I think that a `--self-signed` makes sense for the SAML case (or for a "I just want a quick key/certificate that I can set on my local node and I know what I'm doing" use case) but even if we don't add it, I expect that the users needing it might not depend on certutil to create their signing/encryption key pairs either way - or they will be ok issuing two certutil commands.

**Comment by ywangd** [comment]
>>  But even with that, it could still cause confusion in a TLS context, since our docs differentiate between cert and CA most of the times (if not all the time)

> Can you explain your concern?

What I mean is that our docs talk about separate CA and cert in most places (if not all places). When configuring TLS, the examples have different values (files) for `certificate_authorities` and `certificate` (except for p12 truststore where it's an all-in-one). Self-signed certificate will be a new thing for our docs. I was wondering whether this could be confusing users because 1) `certificate_authorities` and `certificate` will have the same value when configuring TLS and 2) if we do want to explain it, feels like we will be adding docs just for something that is rarely recommended. These are minor concerns. I am actually not too worried about them. Just raise it as discussion point.

Based on the discussion, I think a new `--self-signed` option is the way to go. I think it makes sense for it to be part of this PR. Will work on it.

**Comment by ywangd** [comment]
@tvernum @lcawl @jkakavas The latest push added the `--self-signed` option as discussed, also updated relevant docs. Could you please take another look? Thanks!

File: docs/reference/commands/certutil.asciidoc [hunk 1]
```diff
@@ -82,6 +82,12 @@ All certificates that are generated by this command are signed by a CA. You can
 provide your own CA with the `--ca` or `--ca-cert` parameters. Otherwise, the
 command automatically generates a new CA for you. For more information about
 generating a CA, see the <<certutil-ca,CA mode of this command>>.
+You can also generate self-signed certificates with the `--self-signed` parameters.
+
+deprecated:[7.11.0]
+Generating certificates without specifying a CA certificate is deprecated.
```
Comment (jkakavas) [inline docs/reference/commands/certutil.asciidoc:88]:
```suggestion
Generating certificates without specifying a CA certificate and key is deprecated.
```

File: docs/reference/commands/certutil.asciidoc [hunk 2]
```diff
@@ -82,6 +82,12 @@ All certificates that are generated by this command are signed by a CA. You can
...
+Next major version will mandate a CA certificate to be provided with the
+`--ca` or `--ca-cert` parameters, unless the `--self-signed` option is specified.
```
Comment (jkakavas) [inline docs/reference/commands/certutil.asciidoc:90]:
```suggestion
`--ca` or `--ca-cert` and `--ca-key` parameters, unless the `--self-signed` option is specified.
```

File: x-pack/docs/en/security/authentication/saml-guide.asciidoc [hunk 3]
```diff
@@ -505,11 +505,11 @@ support. Since PEM format is the most commonly supported format, the examples
 below will generate certificates in that format.
 
 Using the <<certutil,`elasticsearch-certutil` tool>>, you can generate a
-signing certificate with the following command:
+self-signed certificate with the following command:
```
Comment (jkakavas) [inline x-pack/docs/en/security/authentication/saml-guide.asciidoc:508]: The most important characteristic is that this is a `signing` certificate (used to sign SAML messages), not that it is a `self-signed` one. I'd suggest we revert this change

File: x-pack/docs/en/security/authentication/saml-guide.asciidoc [hunk 4]
```diff
@@ -505,11 +505,11 @@ support. Since PEM format is the most commonly supported format, the examples
...
 
 [source, sh]
 --------------------------------------------------
-bin/elasticsearch-certutil cert -pem -days 1100 -name saml-sign -out saml-sign.zip
+bin/elasticsearch-certutil cert -self-signed -pem -days 1100 -name saml-sign -out saml-sign.zip
```
Comment (jkakavas) [inline x-pack/docs/en/security/authentication/saml-guide.asciidoc:512]:
```suggestion
bin/elasticsearch-certutil cert --self-signed --pem --days 1100 --name saml-sign --out saml-sign.zip
```

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java [hunk 5]
```diff
@@ -167,8 +167,9 @@ public static void main(String[] args) throws Exception {
 
     static final String CA_EXPLANATION =
         "    * All certificates generated by this tool will be signed by a certificate authority (CA).\n" +
-            "    * The tool can automatically generate a new CA for you, or you can provide your own with the\n" +
-            "         -ca or -ca-cert command line options.";
+            "    * The certificates can be self signed using the -self-signed command line option.\n" +
+            "         Alternatively the tool can automatically generate a new CA for you, or you can " +
+            "         provide your own with the -ca or -ca-cert command line options.";
```
Comment (jkakavas) [inline x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:172]: How about something like: 

```suggestion
"    * All certificates generated by this tool will be signed by a certificate authority (CA), unless" +
            "    the --self-signed command line option is specified.\n" +
            "         The tool can automatically generate a new CA for you, or you can " +
            "         provide your own with the -ca or -ca-cert command line options.";
```

I think it reads better but I'll let Lisa tell us :)

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java [hunk 6]
```diff
@@ -807,7 +819,8 @@ void generateAndWriteSignedCertificates(Path output, boolean writeZipFile, Optio
                         } else {
                             final String fileName = entryBase + ".p12";
                             outputStream.putNextEntry(new ZipEntry(fileName));
-                            writePkcs12(fileName, outputStream, certificateInformation.name.originalName, pair, caInfo.certAndKey.cert,
+                            writePkcs12(fileName, outputStream, certificateInformation.name.originalName, pair,
```
Comment (jkakavas) [inline x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:822]: I also _think_ we don't need to add a certificate entry too when we create a self-signed PKCS12, but did we check?

**Review by jkakavas — COMMENTED** [review]
Just a few comments and suggestions. Can we also update the PR description to mention the self-signed option?

**Review by tvernum — COMMENTED** [review]
Sorry, It looks like I had these comments pending for a week or so & didn't submit them...

File: x-pack/docs/en/security/authentication/saml-guide.asciidoc (hunk 3 above)
Comment (ywangd) [reply x-pack/docs/en/security/authentication/saml-guide.asciidoc]: You are right. Reverted.

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java (hunk 6 above)
Comment (ywangd) [reply x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:822]: This is a good point and it turns out the check is missing ...
I now [augumented](https://github.com/elastic/elasticsearch/pull/64037/files#diff-f63090ebeefa52865c05fb313b5d7e73fd81a4c37f338a7a328df9c74b6134b7R634) `testCreateCaAndMultipleInstances` to test p12 generation for self-signed cert.

File: docs/reference/commands/certutil.asciidoc [hunk 7]
```diff
@@ -211,6 +217,9 @@ wish to password-protect your PEM keys, then do not specify
 `--pem`:: Generates certificates and keys in PEM format instead of PKCS#12. This
 parameter cannot be used with the `csr` parameter.
 
+`--self-signed`:: Generates self-signed certificates. This parameter is only
+applicable to the `cert` parameter.
```
[the remark this answers is not in the document]
Comment (ywangd) [reply docs/reference/commands/certutil.asciidoc]: I added a [note](https://github.com/elastic/elasticsearch/pull/64037/files#diff-efc04dbb0605a05200c191c43916d24c63cc20b81ee797563fadbf0e904b7fbaR222-R225) to state this option should be used with care.

File: x-pack/docs/en/security/authentication/saml-guide.asciidoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/docs/en/security/authentication/saml-guide.asciidoc]: You are right. Ioannis said the same. It's reverted. Thanks.

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java [hunk 8]
```diff
@@ -167,8 +167,9 @@ public static void main(String[] args) throws Exception {
 
     static final String CA_EXPLANATION =
         "    * All certificates generated by this tool will be signed by a certificate authority (CA).\n" +
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java]: Ioannis has kindly offered suggestion and it should be accurate now.

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java [hunk 9]
```diff
@@ -330,6 +331,9 @@ CAInfo getCAInfo(Terminal terminal, OptionSet options, Environment env) throws E
             } else if (options.has(caCertPathSpec)) {
                 return loadPemCA(terminal, options, env);
             } else {
+                terminal.println("Note: Generating certificates without providing a CA certificate is deprecated.");
+                terminal.println("      A CA certificate will become mandatory in next major release.");
```
Comment (jkakavas) [inline x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:335]:
```suggestion
                terminal.println("      A CA certificate will become mandatory in the next major release.");
```

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java (hunk 6 above)
Comment (jkakavas) [reply x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:822]: My point was actually about the construction of the PKCS#12 bundle and whether or not it suffices to add a private key entry with the key and certificate, or if we also need to add a trusted certificate entry with the certificate too. I believe we don't and other tooling ( keytool, openssl ) seem to agree too :)

**Review by jkakavas — APPROVED** [review]
LGTM, I'll defer more scrutiny of the wording changes to Lisa and Tim

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java [hunk 10]
```diff
@@ -166,9 +166,10 @@ public static void main(String[] args) throws Exception {
             "      disable hostname verification in your SSL configuration.";
 
     static final String CA_EXPLANATION =
-        "    * All certificates generated by this tool will be signed by a certificate authority (CA).\n" +
-            "    * The tool can automatically generate a new CA for you, or you can provide your own with the\n" +
-            "         -ca or -ca-cert command line options.";
+        "    * All certificates generated by this tool will be signed by a certificate authority (CA)\n" +
+            "      unless the --self-signed command line option is specified.\n" +
+            "      The tool can automatically generate a new CA for you, or you can provide your own with " +
+            "      the -ca or -ca-cert command line options.";
```
Comment (lcawl) [inline x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:172]: Should these have two dashes?
```suggestion
            "      the --ca or --ca-cert command line options.";
```

File: docs/reference/commands/certutil.asciidoc [hunk 11]
```diff
@@ -73,15 +73,18 @@ directory name, you must also specify a file name in the `--name` command
...
 about this setting, see <<security-settings>>.
 
-All certificates that are generated by this command are signed by a CA. You can
-provide your own CA with the `--ca` or `--ca-cert` parameters. Otherwise, the
-command automatically generates a new CA for you. For more information about
-generating a CA, see the <<certutil-ca,CA mode of this command>>.
+All certificates that are generated by this command are signed by a CA unless
+the `--self-signed` parameter is specified. You can provide your own CA with the
+`--ca` or `--ca-cert` and `--ca-key` parameters. Otherwise, the command automatically generates a new CA for you.
+deprecated:[7.11.0,"Generating certificates without specifying a CA certificate and key is deprecated. In the next major version you must provide a CA certificate unless the `--self-signed` option is specified."]
```
Comment (lcawl) [inline docs/reference/commands/certutil.asciidoc:84]: > deprecated:[7.11.0,"Generating certificates without specifying a CA certificate and key is deprecated. In the next major version you must provide a CA certificate unless the `--self-signed` option is specified."]

I've combined this information into a single paragraph so that it's (hopefully) clearer what's deprecated.

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:172]: We haven't really been very consistent on the usage of single dash and double dashes. There are still many other places that use single dash. For this particular paragraph, I am ok to make it all double dashes.

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java [hunk 12]
```diff
@@ -818,17 +831,26 @@ void generateAndWriteSignedCertificates(Path output, boolean writeZipFile, Optio
...
+            if (caInfo != null) {
+                certificate = CertGenUtils.generateSignedCertificate(certificateInformation.name.x500Principal,
+                    getSubjectAlternativeNamesValue(certificateInformation.ipAddresses, certificateInformation.dnsNames,
+                        certificateInformation.commonNames),
+                    keyPair, caInfo.certAndKey.cert, caInfo.certAndKey.key, days);
+            } else {
+                certificate = CertGenUtils.generateSignedCertificate(certificateInformation.name.x500Principal,
+                    getSubjectAlternativeNamesValue(certificateInformation.ipAddresses, certificateInformation.dnsNames,
+                        certificateInformation.commonNames),
+                    keyPair, null, null, true, days, null);
```
Comment (tvernum) [inline x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java:852]:
```suggestion
                    keyPair, null, null, false, days, null);
```
The `true` means that this is a CA, but then `--self-signed` is just doing the same thing as `certutil ca`, but it should be generating a non ca, self-signed cert,

File: x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java [hunk 13]
```diff
@@ -645,9 +649,12 @@ void generateAndWriteCsrs(Path output, int keySize, Collection<CertificateInform
 
     static class GenerateCertificateCommand extends CertificateCommand {
 
+        OptionSpec<Void> selfSigned;
+
         GenerateCertificateCommand() {
             super("generate X.509 certificates and keys");
             acceptCertificateGenerationOptions();
+            selfSigned = parser.accepts("self-signed", "generate self signed certificates (takes precedence over the ca options)");
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java]: Good catch, Thanks. I misunderstood the existing behaviour of `--ca` and `--ca-cert`. The two plus the new `--self-signed` should be mutually exclusive. It is now [fixed](https://github.com/elastic/elasticsearch/pull/64037/files#diff-766b87da83658942494367bd114ab7a9c239821ca858cabc0673d41da5ca9c4aR660-R661).

**Review by tvernum — APPROVED** [review]
LGTM
